dbus/dbus.go

```diff
@@ -0,0 +1,31 @@
+// Copyright 2020 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package dbus
+
+import "github.com/pkg/errors"
+
+var dbusInterface DBusInterface = nil
+
+type DBusInterface interface {
```

Comment:
Question:
I don't fully understand why the interface is needed here. Can't the package just define the interface?
i.e:
```
dbus.GenerateGUID()
...
```

Reply:
@oleorhagen If the package defines the interface here (in the always compiled-in file), were do we put the implementation (i.e. the maybe opt-out code)?

Reply:
Just playing with ideas here, but something along the lines of:

```
diff --git a/dbus/dbus.go b/dbus/dbus.go
index 41bdbf7..de4b020 100644
--- a/dbus/dbus.go
+++ b/dbus/dbus.go
@@ -16,16 +16,3 @@ package dbus
 
 import "github.com/pkg/errors"
 
-var dbusInterface DBusInterface = nil
-
-type DBusInterface interface {
-	GenerateGUID() string
-	IsGUID(string) bool
-}
-
-func GetDBusInterface() (DBusInterface, error) {
-	if dbusInterface != nil {
-		return dbusInterface, nil
-	}
-	return nil, errors.New("no DBus interface available")
-}
diff --git a/dbus/dbus_libgio.go b/dbus/dbus_libgio.go
index 7ab000e..44efbc3 100644
--- a/dbus/dbus_libgio.go
+++ b/dbus/dbus_libgio.go
@@ -19,8 +19,8 @@ package dbus
 // #include <gio/gio.h>
 import "C"
 
-type DBusInterfaceLibGio struct {
-	d DBusInterface
+func Available() error {
+	return nil
 }
 
 // GenerateGUID generates a D-Bus GUID that can be used with e.g. g_dbus_connection_new().
diff --git a/dbus/nodbus_libgio.go b/dbus/nodbus_libgio.go
new file mode 100644
index 0000000..2cddaf0
--- /dev/null
+++ b/dbus/nodbus_libgio.go
@@ -0,0 +1,16 @@
+
+// +build nodbus,cgo
+
+package dbus
+
+func Available() error () {
+	return errors.New("No D-Bus available")
+}
+
+func GenerateGUID() string {
+	return ""
+}
+
+func IsGUID(str string) bool {
+	return false
+}
```

Reply:
I prefer Lluís' method, otherwise we have to define every function we don't use.

Reply:
Yes, it is just a suggestion :)

You're right, they have to be defined. However, the readability is pretty good :) It's a trade-off.

Reply:
@oleorhagen Now I understand what you meant.

But as Kristian also said I think we would need a lot of boiler plate code. Besides, these are the trivial methods, but we most likely require others with custom structs as parameters and it will get too ugly to mock all of that.